Macro Editor

algernon opened the pull request:
This is a _very_ rough draft of a macro editor. It doesn't do much yet, it can only display macros, but doesn't allow any editing at all yet.

However, the underlying `api/macros` stuff does support serializing, it's just the UI parts that are missing. I mostly present this to ask for feedback about the direction of the UI, which currently looks like this:

![Screenshot from 2021-08-02 13-54-56](https://user-images.githubusercontent.com/17243/127858210-4030064c-332a-4e4a-a6bb-40bc531ad2ca.png)

The idea is to present each macro as a card, with the steps as a list. Eventually, the steps will be draggable to allow easy reordering, and they'll have a trashcan button to remove a step, too. Clicking on a step will allow changing the values of the step: its type, and its payload. If we're dealing with a step that involves keys, we'll re-use the key selection sidebar we use for the keymap editor. For stuff like interval and wait, we'll allow in-line editing.

Somewhere near the top, we'd also want to display the storage size available for macros, and how much of it we're using. We'll also need a save button at the usual place.

obra commented:
Awesome.


I *think* that rows instead of cols might make more sense?

algernon commented:
That can be arranged, yes, and it does make for a nicer layout.

jongalloway commented:
I've been watching this PR since it was opened, can't wait for it. I've tried running it locally to contribute, but it just hangs on the Macro Editor screen after connecting to my keyboard, both Mac and Windows. An issue with the branch?

gedankenexperimenter commented:
> …select the type (key up, key down, tap, delay, interval, send_report - only those, initially)…

I want to discourage including `send_report`.  It's not really very useful anymore, now that Macros are generating `KeyEvent`s rather than fiddling with the Keyboard HID report.

algernon commented:
> > …select the type (key up, key down, tap, delay, interval, send_report - only those, initially)…
> 
> I want to discourage including `send_report`. It's not really very useful anymore, now that Macros are generating `KeyEvent`s rather than fiddling with the Keyboard HID report.

Hmm. Upon going over the scenarios I had in mind where `send_report` would be useful, it turns out none of them really need it. So you're correct, we should not offer `send_report` on the UI, either.

algernon commented:
Current progress:

![Screenshot from 2022-05-21 14-10-11](https://user-images.githubusercontent.com/17243/169650906-57b8d151-418d-4aa7-811e-89ed03e15739.png)

There is no editing functionality yet, this is purely display, and rather limited at that, too. There's no support for the keycode variants, nor do we handle the likes of `explicit_report`, `implicit_report` or `send_report`. To test it, we'll need to manually set a dynamic macro, for example with the `focus-test` tool shipped with Kaleidoscope:

```
$ focus-test macros.map 5 0 4 5 0 5 0 1 20 12 0 4 0 5 0 6 0 0 2 128 5 0 4 3 0 225 12 0 27 0 28 0 29 0 0 4 0 225 0
```

This sets up two macros, the second of which is shown in the screenshot.

## Observations

I think that with the floating key picker, our best option here is to replace the keymap with the macro editor, rather than popping up a dialog or popover or something. That would also let us use the application context to cancel any unsaved macro edits. Technically, we can do that with a popover too, but... replacing the keymap with the macro editor feels cleaner.

As far as icons go, I'm not entirely happy about the keyup/keydown icons. It's tempting to use the Expand / Compress material icons instead, those might better convey visually what the step does.

## Scope

Before going wild with ideas, I'd like to set the scope of the initial macro support: I want to support the `wait`, `interval`, `keydown`, `keyup`, `tap`, and `tapsequence` steps, and nothing else. Anything else should be treated as opaque, something that can be deleted only (see below for more info). There will be no macro optimization on save, either - but that can and will be added later.

There will not be any step rearrangement features yet, not in the initial version. We'll be able to delete existing steps, and add new ones, _perhaps_ edit existing ones too, but add & delete are the core functionality, anything else I consider nice to have, but non-essential for the initial version.

## Ideas

We can display unhandled macro steps with a `?` icon, and the human-friendly step name in the chip label. We wouldn't be able to edit those, but we could delete them and rearrange them (once we implement drag & drop in a future iteration).

We will want to display the available macro space, and how much of it we use.

We will want to make it possible to delete a macro. That presents a few problems, however: macros are indexed. If we simply delete one, any other macro will shift. We should be careful with that, but should allow doing so. Other than that, we can have empty macros, that only takes a single byte. That's something the current display code probably doesn't handle.

algernon commented:
Another observation: The `DynamicMacros` plugin uses different keycodes than `Macros`. I'm not convinced anymore that this is a good idea. We should do something similar to what `DynamicTapDance` does: use the same keycodes, and rely on the `macroAction` handler to call into us as needed:

```
const macro_t *macroAction(uint8_t macro_id, KeyEvent &event) {
  if (macro_id == 0) return runMacro0(event);
  return DynamicMacros.play(macro_id);
};
```

The major disadvantage of having different keycodes is complexity. We removed support for dealing with hardcoded layers from Chrysalis, because the hardcoded/custom distinction was confusing, and made a lot of workflows fragile. Same applies to Macros & DynamicMacros. From an end-user point of view, there are macros. Where they are stored, should not matter.

So I propose we change the `DynamicMacros` plugin to be more like `DynamicTapDance`, and use the same codes as normal macros. That reduces cognitive load, makes the UI simpler too. We will need to know how many built-in macros there are, though, so that the macro editor can properly say "I'm sorry $Dave, I can't do that".

Another advantage of reusing Macro keycodes is that we can have a total of 256 macros, all editable, if need be, while with the current DynamicMacros setup, we can only have 32 editable ones (on top of the 256 built-in ones). So while we'd have less available macros in total, we'd have _much_ more dynamic macros.

obra commented on src/api/macros/chrysalis-macros.js:
```diff
@@ -249,15 +218,64 @@ class Macros {
+  const compressMacro = (macro) => {
+    let m = clone(macro);
+    for (let step of m) {
+      if (step.type == "KEYUP" && step.value.code < 256)
```
step.value.code < 256 feels like it's something that could be an "isFoo" method on step.value

obra commented on src/api/macros/chrysalis-macros.js:
```diff
@@ -249,15 +218,64 @@ class Macros {
-      if (ser.length > macros.storageSize) {
+      const m = this.compress(macros);
+      const ser = this.serialize(m.macros);
+      if (ser.length > m.storageSize) {
```
Eventually, we should have a friendlier error message. At least "Your macros take x bytes, but your device only has y bytes for macro storage. "

algernon replied:
The message here isn't really interesting. There's `getStoredSize()` which the UI will use to check the size, and will disallow saving in the first place, so the UI will never hit this path. It's only there for development and possible reuse outside of the UI itself.

obra replied:
Did that get sorted?

algernon replied:
Yes, the "Save changes" button on the main editor page is disabled when the macro space is overused, and a warning is displayed:

![Screenshot from 2022-05-27 22-24-37](https://user-images.githubusercontent.com/17243/170784672-50ae5ca3-3dba-4179-aa7e-e19ca74719a8.png)

obra commented:
I agree that macros and dynamicmacros really ought to use the same keycodes.

I think that at least for now icons + tooltips is going to be worse UX than just using words:

[ Delay between steps: 20ms] [ Tap "A"] [Tap "B"] [Tap "C"] [Wait 128ms] [Press "Shift"] [Tap "X"] [Tap "Y"] [Tap "Z"] [Release "Shift"]

Also, doing this with text on screen will be quicker to iterate on.  I'd much rather we get to editable macros sooner than figure out a gorgeous representation of macro steps.

obra commented:
>I think that with the floating key picker, our best option here is to replace the keymap with the macro editor, rather than popping up a dialog or popover or something. That would also let us use the application context to cancel any unsaved macro edits. Technically, we can do that with a popover too, but... replacing the keymap with the macro editor feels cleaner.

Except I think we're going to want the keypicker in the macro editor pretty soon. I wonder if we want "Edit one macro" to be a screen rather than a widget. I'd previously been concerned about having an "Edit all the macros" screen as a thing, but doing it for the edit experience for a single macro might not be wrong.

obra commented:
> We will want to make it possible to delete a macro. That presents a few problems, however: macros are indexed. If we simply delete one, any other macro will shift. We should be careful with that, but should allow doing so. Other than that, we can have empty macros, that only takes a single byte. That's something the current display code probably doesn't handle.

Yeah, I think that shifting macro positions is out of scope for v1.

algernon commented:
> I agree that macros and dynamicmacros really ought to use the same keycodes.

I'll make that happen then. There are some details to be discussed, I'll note those down in keyboardio/Kaleidoscope#1169.

> I think that at least for now icons + tooltips is going to be worse UX than just using words:
> 
> [ Delay between steps: 20ms] [ Tap "A"] [Tap "B"] [Tap "C"] [Wait 128ms] [Press "Shift"] [Tap "X"] [Tap "Y"] [Tap "Z"] [Release "Shift"]

Can do that, too. The step rendering is neatly in its own component, so I'll just replace it with a textual thing. The main reason I went with icons + tooltips + text is because that was easier to implement, because I had all the components ready. But... I suppose I can just replace the label with `<strong>Tap:</strong> A B C`. I'll give that a try in a bit.

> Except I think we're going to want the keypicker in the macro editor pretty soon. I wonder if we want "Edit one macro" to be a screen rather than a widget. I'd previously been concerned about having an "Edit all the macros" screen as a thing, but doing it for the edit experience for a single macro might not be wrong.

Oh, we'll definitely keep the key picker! I meant replacing the keyboard image with the single-macro editor, and keeping the sidebar and the key picker.

So it will _feel_ like a screen, but will be embedded into the Editor.

obra commented:
> Oh, we'll definitely keep the key picker! I meant replacing the keyboard image with the single-macro editor, and keeping the sidebar and the key picker.

Oh. You mean the replacing image of the user's device. not the 104key. got it.

algernon commented:
![Screenshot from 2022-05-21 22-18-42](https://user-images.githubusercontent.com/17243/169667754-a25c015b-00a2-469c-ae4c-d6987614c800.png)

The pushed code doesn't quite look like this, because I manually added a `flex-wrap: wrap` property for the sake of the screenshot. But this is the general idea about how I want it to look.

algernon commented:
Something to ponder on: how should we trigger the macro editor? For now, if we select a macro key, it pops up automatically. That's great for development, but not for normal use.

We could have an "Edit current macro" button on the sidebar, which greys out if we're not focusing a macro button. Or we could have a right-click context menu on the buttons, which, for macro buttons, would show an "Edit..." option. For any other button, it'd just show a "Close" option, or not pop up anything at all.

The context menu I feel is nicer, but it's also considerably harder to implement. So for v1, I'd go with the button on the sidebar, and we can improve on that later.

algernon commented:
Oh, and a quick progress report:

![Screenshot from 2022-05-22 00-04-53](https://user-images.githubusercontent.com/17243/169670384-583c3177-d8f3-47c4-b3ef-be2cfd5e71cc.png)

The delete buttons on the chips **do** remove the appropriate step now. We can't save the results yet, but we do remove the step.

The editor keeps an internal copy of the macro, which will not be persisted until the "Apply" button is hit. So we can freely poke around in the editor, hit the close button, and nothing will be touched. Or we can press Apply, and our changes will be propagated to the editor. To save it to the keyboard, we'll still need to hit the save button in the Editor too...

Or, perhaps we shouldn't have two steps. Perhaps "Apply" could save it to the keyboard immediately, skipping a step. We do not want to reuse the "Save Changes" button, because that's for saving the keymap state. We should probably hide that button when we have the macro editor open, to be honest.

algernon commented:
![Screenshot from 2022-05-22 04-14-01](https://user-images.githubusercontent.com/17243/169675387-68a313be-f3b1-40d1-a7b5-7489d3b8dcc5.png)

Getting there, slowly!

I need to go to sleep, so here's a brain dump so I can pick stuff up tomorrow:

We need to move the WIP macro state up to the Editor, unless we want to mess with channels and whatnot, and that'd be ugly. We need to do this, so that `onKeyChange` reaches everything it needs to. In case of the Keymap, we need to update the keymap itself. In case of Macros, we need to update the currently selected step of the currently selected macro. We either broadcast messages from way down, or we move the wip macro state up. Moving the WIP macro state up makes more sense, because then we can use the normal context stuff and the save button, too. We won't have two kinds of saves: one for the keymap, one for macros. We're on the Editor screen. We save both. This has the disadvantage that if we edited the keymap, didn't save, then edit macros, if we want to undo the macro edits, we'll undo the keymap changes too. That's an acceptable sacrifice, imo.

This in turn means that we will have to turn TAPSEQUENCE into discrete TAPs, and compress subsequent TAPs back into a TAPSEQUENCE on save. Otherwise we'd have too many variables and state to juggle around.

Right. That's about editing.

We'll need a few more things for a minimally functional macro editor:

- We need to be able to add new steps.
- We need to be able to "add" new macros.

Right now, when we try to edit a macro that doesn't exist on the keyboard, we get a blank screen. When we try to edit a macro that doesn't exist yet, we need to fill up our macro set to have macros up to the desired index.

We also need to teach the whole thing about empty macros.

Oh, and we need to make sure that when we save, we have enough space. We should display the current macro space usage somewhere.

algernon commented:
We could display the available macro space on the sidebar. Either in the macro section, or below the layers, to be slightly easier to spot.

algernon commented:
Another idea: Instead of having a separate StepEditor card, since the "cancel" and "apply" buttons are going away from the macro step card, we could have the edit widget right below the macro steps, with a divider inbetween. The main card will have its Close icon back on the top.

For the new step button: it could pop up a menu, right at the cursor, to select the type of step, and then we'd simply add a new step with reasonable default values, and immediately select it.

algernon commented:
Time for another update!

![Screenshot from 2022-05-22 16-03-32](https://user-images.githubusercontent.com/17243/169699085-0a211dd8-953b-4aec-a731-99a0392dc615.png)

We can now:
- Edit macros: remove steps, add steps, change steps (their value, not their type)
- "Create" new macros: just set a key to a specific macro key, hit the edit button, and off you go!
- Use the Floating Key Picker to change press/release/tap key values.
- Save macros back to the keyboard
- Glance at the sidebar and see how much macro space we have left

There's a whole lot of things under the hood, and the whole thing is a bit fragile, and a couple of important features are still missing.

Known issues:
- There is no warning when we go over the macro storage limit. The exception from `@api/macros` just bubbles up and we fall over.
- Pretty much everything on the sidebar is broken with regards to the macro editor.
- We still need to tackle keyboardio/Kaleidoscope#1169 before this really becomes usable.

algernon commented:
Couple of notes, while I'm cleaning up Sidebar stuff:

We should be using `Input` directly, rather than `TextField` for numeric input in the INTERVAL & WAIT editors, so that we can have validation.

algernon commented:
Update: Everything on the sidebar works, and we do not require any changes on the Kaleidoscope part.

We still do not have a warning when we overflow the storage, but I'll get there soon.

algernon commented:
Now we have a notification too, and things appear to be working as expected when connected to a device without dynamic macro support.

@obra I think think this is ready for an initial review.

I will still collapse the branch history, but I think that at this time, the current commits make it easier to review, as they have useful info here and there. So not taking the Draft status off just  yet.

algernon commented:
One thing that IS missing is properly handling unknown step types. In case of corruption, or setting up the macros outside of Chrysalis, these may happen. We currently display them, as `<unkown>`, but we could do better than that. A chip with a warning sign, and `**Unsupported:** $step` (for known, but unsupported steps; `$step` would still remain `unknown` for really unknown steps) could do the trick, I suppose.

algernon commented:
Now it is _really_ ready for review. Not  for merge yet (I'll clean up the history before that), but for review.

algernon commented:
@obra What should I do about the branch history?

On one hand, 74 commits are a lot, and a lot of it is just figuring things out. On the other hand, it's an accurate representation of the development story of the feature (with only 2 wip commits I can easily reword). I can squash it all down to one massive commit to make the history cleaner, but we lose a lot of information about the why things ended up being the way they are in that case.

In either case, I don't have strong feelings either way. I'd prefer a full history over a squash, but I'm not opposed to squash it down to one commit either, if you think that's preferable.

obra commented on src/api/macros/chrysalis-macros.js:
```diff
@@ -0,0 +1,386 @@
+/* Chrysalis -- Kaleidoscope Command Center
```
The flow from here to macros/index.js to App.js seems confusing to me. 
Is there a reason this file shouldn't just be src/api/macros.js?

algernon replied:
This could be `api/macros/index.js` itself. The reason it's not is because I originally copied another module that had a similar setup, and never bothered to change it.

I can certainly move it all to a single `src/api/macros/index.js`, or even `src/api/macros.js`. The former would be more consistent with the rest of the modules, some of which are similarly index.js + random other (focus, log, colormap are like that).

obra replied:
I'm not a big fan of the index.js pattern, both because it's kind of nonsensical and because then all my editor tabs are for 'index.js'

algernon replied:
Hmm.... there's only 3 `index.js` files in `src/api/focus` currently, but... I suppose I can cut that down to 0.

obra commented on src/renderer/App.js:
```diff
@@ -18,6 +18,7 @@
 import "@api/colormap";
 import Focus from "@api/focus";
 import "@api/keymap";
+import "@api/macros";
```
I suspect that this import is to get the macros commands registered in Focus? is that right? Given that we always want to do that, why not just have @api/focus do that import?

algernon replied:
Historical reasons, mostly, but they no longer apply, really. Would make sense to merge `colormap` and `macros` into focus, indeed.

obra commented on src/renderer/screens/Editor/Editor.js:
```diff
@@ -257,36 +336,67 @@ const Editor = (props) => {
+    selectorKey || keymap.custom[currentLayer][currentKeyIndex];
+
+  let mainWidget;
+  if (openMacroEditor) {
```
Toggling the main widget like this feels kind of weird to me, but I can't put my finger on why.

algernon replied:
It certainly isn't the cleanest thing. For now, I'd leave it as-is, and when I do a Suspense + Skeleton pass over all of Chrysalis, I'll clean this (and similar patterns elsewhere, too) up. The whole of Editor could use some refactoring, but I feel that's best done separately.

obra replied:
ok

obra commented on src/renderer/screens/Editor/Macros/MacroEditor.js:
```diff
@@ -0,0 +1,156 @@
+import MacroStepAdd from "./MacroStepAdd";
+import MacroStepEditor from "./MacroStepEditor";
+
+const focus = new Focus();
```
Do we use this?

algernon replied:
We do not, good catch. Thanks! We don't use KeymapDB + db, either.

obra commented on src/renderer/screens/Editor/Macros/MacroEditor.js:
```diff
@@ -0,0 +1,156 @@
+  const { t } = useTranslation();
+
+  const onStepDelete = (index) => {
+    const m = macro.map((v) => Object.assign({}, v));
```
Where do we get lowercase `macro` from? I don't think we likely want to be throwing something with quite so generic a name into our global namespace without an explicit import. (And also, not being able to track it down in code will be..frustrating for folks working on the codebase.)

We might also want to think about renaming this thing to something like KaleidoscopeMacro or ..somethign?

algernon replied:
> Where do we get lowercase `macro` from? 

A few lines up:

`const { macroId, macro, onMacroChange, macroStep, setMacroStep } = props;`

From `props`.

> I don't think we likely want to be throwing something with quite so generic a name into our global namespace

We're not, it's coming from props. We could use a better name for either `m` or `macro`, however. `newMacro` instead of `m` sounds fine to me. The purpose of `m` is to clone the current macro, adjust it according to the edit, and then replace the original with it. We need to clone it, because otherwise we end up editing the original, and React's state diffing won't kick in.

obra replied:
I just missed it in props. Thank you.

obra commented on src/renderer/screens/Editor/Macros/MacroStep.js:
```diff
@@ -0,0 +1,88 @@
+  };
+
+  const formatStep = (step) => {
+    if (step.type == "INTERVAL" || step.type == "WAIT")
```
Is there a reason we want macro step types to be strings rather than constants we can compare against?

algernon replied:
The only reason is that we also use the strings for translations, like: `return t("editor.macros.steps." + step.type);` a few lines below.

We could, of course, have a function to map constants to strings or something along those lines, but that feels more complicated. More correct, but also more complicated. Is the added complexity worth it?

obra replied:
The types could be string constants, which really just means we'd need to define them somewhere. This mostly feels like a "not hurting ourselves down the road" thing.

algernon replied:
> [61359c0](https://github.com/keyboardio/Chrysalis/commit/61359c0f56cff3aa280701610b3f8fde9d29bf90) looks like a nice improvement, but I'm not entirely sure I understand the "why" of the changes from the commit message.

ACK. I'll update the commit message.

The goal is to let us use the sidebar both for changing both keys on the keymap, and the keys used during a macro, with the same sidebar. We're passing `props.onKeyChange` to the Sidebar already, so this PR modifies `onKeyChange`: if the macro editor is open, key change is handled one way, if it isn't, it's handled another way. For this to work, I needed to stop the Sidebar from doing layer + index-based lookups instead (as that doesn't make sense for macro steps), and accept a `props.currentKey` property instead, to let the Editor control what the Sidebar sees, rather than both doing the lookups themselves.

obra reviewed (commented):
61359c0f56cff3aa280701610b3f8fde9d29bf90 looks like a nice improvement, but I'm not entirely sure I understand the "why" of the changes from the commit message.

obra commented:
(Let's keep the history)

On Fri, May 27, 2022 at 6:04 AM Gergely Nagy ***@***.***>
wrote:

> @obra <https://github.com/obra> What should I do about the branch history?
>
> On one hand, 74 commits are a lot, and a lot of it is just figuring things
> out. On the other hand, it's an accurate representation of the development
> story of the feature (with only 2 wip commits I can easily reword). I can
> squash it all down to one massive commit to make the history cleaner, but
> we lose a lot of information about the why things ended up being the way
> they are in that case.
>
> In either case, I don't have strong feelings either way. I'd prefer a full
> history over a squash, but I'm not opposed to squash it down to one commit
> either, if you think that's preferable.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Chrysalis/pull/703#issuecomment-1139599472>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2AYKPVFW4MM6T3GTLDVMDB5FANCNFSM5BMSXPJQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

algernon commented:
A couple of issues I noticed while replying to review notes:

- When I have a non-TAP step selected, the sidebar still works, and will happily try to assign a Key object to a WAIT step, which will not end well. If we're non a non-TAP step, we should probably disable the sidebar.
- When adding a WAIT/INTERVAL step, the input box sometimes doesn't pop up. It only pops if I had another step selected: if I had none, then the box doesn't appear until I re-select the step.

algernon commented:
Addressed all the review comments, and rebased on top of master. I also adjusted the commits that needed rewording: the wip commits, and the highlighted one: they all have adequate commit messages now.

I still need to address the [two issues I found](https://github.com/keyboardio/Chrysalis/pull/703#issuecomment-1140031879) earlier, however.

algernon commented:
The two bugs mentioned above are now fixed too.

obra commented:
I was figuring that focus was going to end up at focus.js and the others were going to end up at focus/blah.js. "focus/focus.js" feels wrong.

algernon commented:
> I was figuring that focus was going to end up at focus.js and the others were going to end up at focus/blah.js. "focus/focus.js" feels wrong.

Understood. I'll fix that in a bit.
